src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs

```diff
@@ -56,22 +78,74 @@ private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitializatio
                     _builder.EmitToken(field, inits.Syntax, _diagnostics.DiagnosticBag);
                     _builder.EmitIntConstant(data.Length);
                     _builder.EmitOpCode(ILOpCode.Cpblk, -3);
-
-                    if (initializationStyle == ArrayInitializerStyle.Mixed)
+                }
+                else
+                {
+                    if (getCreateSpanHelper(_module, elementType) is { } createSpanHelper &&
+                        _module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__GetPinnableReference) is MethodSymbol getPinnableReference)
                     {
-                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
+                        // Use RuntimeHelpers.CreateSpan and cpblk.
+                        EmitStackAllocBlockMultiByteInitializer(data, createSpanHelper, getPinnableReference, elementType, inits.Syntax, _diagnostics.DiagnosticBag);
+                    }
+                    else
+                    {
+                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: true);
+                        mixedInitialized = true;
                     }
                 }
-                else
+
+                if (initializationStyle == ArrayInitializerStyle.Mixed && !mixedInitialized)
                 {
-                    EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: true);
+                    EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
                 }
             }
+
+            void emitLocalloc()
+            {
+                EmitExpression(count, used: true);
+
+                _sawStackalloc = true;
+                _builder.EmitOpCode(ILOpCode.Localloc);
+            }
+
+            static Cci.IMethodReference? getCreateSpanHelper(Emit.PEModuleBuilder module, TypeSymbol elementType)
+            {
+                var member = module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
```

Comment:
>module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle); [](http://example.com/codeflow?start=29&length=136)

The fact that we can use helper should be checked by using a special helper. A null check is not sufficient.

Reply:
We have an example in EmitArrayInitializer.cs